src/cli_args.zig

```diff
@@ -0,0 +1,851 @@
+//! Command line argument parsing for the CLI
+const std = @import("std");
+const testing = std.testing;
+const mem = std.mem;
+const collections = @import("collections.zig");
+const exitOnOom = collections.utils.exitOnOom;
+
+/// The core type representing a parsed command
+/// We could use anonymous structs for the argument types instead of defining one for each command to be more concise,
+/// but defining a struct per command means that we can easily take that type and pass it into the function that implements each command.
+pub const CliArgs = union(enum) {
+    run: RunArgs,
+    check: CheckArgs,
+    build: BuildArgs,
+    format: FormatArgs,
+    test_cmd: TestArgs,
+    repl,
+    version,
+    docs: DocsArgs,
+    help: []const u8,
+    licenses,
+    problem: CliProblem,
+
+    pub fn deinit(self: CliArgs, gpa: mem.Allocator) void {
+        switch (self) {
+            .format => |fmt| gpa.free(fmt.paths),
+            .run => |run| gpa.free(run.app_args),
+            else => return,
+        }
+    }
+};
+
+/// Errors that can occur due to bad input while parsing the arguments
+pub const CliProblem = union(enum) {
+    missing_flag_value: struct {
+        flag: []const u8,
+    },
+    unexpected_argument: struct { cmd: []const u8, arg: []const u8 },
+    invalid_flag_value: struct {
+        value: []const u8,
+        flag: []const u8,
+        valid_options: []const u8,
+    },
+};
+
+/// The optimization strategy for the compilation of a Roc program
+pub const OptLevel = enum {
+    size, // binary size
+    speed, // binary speed
+    dev, // speed of compilation
+
+    fn from_str(str: []const u8) ?OptLevel {
+        if (mem.eql(u8, str, "speed")) return .speed;
+        if (mem.eql(u8, str, "size")) return .size;
+        if (mem.eql(u8, str, "dev")) return .dev;
+        return null;
+    }
+};
+
+/// Arguments for the default `roc` command
+pub const RunArgs = struct {
+    path: []const u8,
+    opt: OptLevel = .dev,
+    app_args: []const []const u8 = &[_][]const u8{},
+};
+
+/// Arguments for `roc check`
+pub const CheckArgs = struct {
+    path: []const u8,
+    main: ?[]const u8,
+};
+
+/// Arguments for `roc build`
+pub const BuildArgs = struct {
+    path: []const u8,
+    opt: OptLevel,
+    output: ?[]const u8 = null,
+};
+
+/// Arguments for `roc test`
+pub const TestArgs = struct { path: []const u8, opt: OptLevel, main: ?[]const u8 };
+
+/// Arguments for `roc format`
+pub const FormatArgs = struct {
+    paths: []const []const u8,
+    stdin: bool = false,
+    check: bool = false,
+};
+
+/// Arguments for `roc docs`
+pub const DocsArgs = struct {
+    path: []const u8,
+    output: []const u8,
+    root_dir: ?[]const u8 = null,
```

Comment:
For these definitions, I'm finding it a little hard to understand at first glance what these are for. It would be helpful to add a comment on things like `path` or `main`, or `root_dir` etc to explain what we are expecting here.